ls1intum/Artemis
# `Tutorial groups`: Allow instructors to import tutorial group registrations form a csv file

### Checklist
#### General


#### Server
#### Client

### Motivation and Context


- This allows Instructors to import tutorial groups and register students to them via import of a .csv file.

### Description

- New endpoint in the server for importing tutorial group registrations
- Client Modal for the import (see screenshots)

### Steps for Testing


For testing I provide you with multiple .csv files that work in the following course on TS1:
https://artemis-test1.artemis.in.tum.de/course-management/62/tutorial-groups-management


[Invalid: Duplicated Logins -> Will show up at parsing](https://github.com/ls1intum/Artemis/files/9668198/invalid_duplicatedLogins.csv)
[Invalid: Duplicated Matriculation Number -> Will show up at parsing](https://github.com/ls1intum/Artemis/files/9668199/invalid_duplicatedRegistrationNumbers.csv)
[Invalid: Missing Tutorial Group Title -> Will show up at parsing](https://github.com/ls1intum/Artemis/files/9668200/invalid_missingTitle.csv)
[Invalid: Unknown User -> Will show up AFTER import](https://github.com/ls1intum/Artemis/files/9668201/invalid_unknownUser.csv)
[Valid: Tutorial Group + Registrations](https://github.com/ls1intum/Artemis/files/9668202/valid_groups_and_registrations.csv)
[Valid: Just Tutorial Groups](https://github.com/ls1intum/Artemis/files/9668203/valid_just_groups.csv)

Prerequisites:

1. Log in to Artemis on TS1
2. Navigate to Course Administration: https://artemis-test1.artemis.in.tum.de/course-management/62/tutorial-groups-management (this course has the users described in the above csvs)
3. Import the above mentioned .csv files and check:
- Tutorial Groups with the titles in the .csv are created if they do not exist yet
- Students will be deregistered from their current tutorial group and registered to the one specified in the .csv
- Invalid CSVs should show an error

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| TutorialGroupService.java | 100% | 100% |
| TutorialGroupResource.java | 96% | 94% |
| tutorial-group-row-buttons.component.ts | 75% | 95% |
| tutorial-groups-registration-import-dialog.component.ts | 76% | 97%



### Screenshots


![image](https://user-images.githubusercontent.com/29718932/192867296-a2c13c36-c645-4535-a620-145c1161358c.png)
![image](https://user-images.githubusercontent.com/29718932/192868829-9bfb9f34-8924-48f4-9ea0-28f328bd4d98.png)
![image](https://user-images.githubusercontent.com/29718932/192869565-c2365851-cf56-499b-93ab-53139b708478.png)
![image](https://user-images.githubusercontent.com/29718932/192869801-f1a2a12f-a4e5-4513-a283-7ee9936e77c1.png)
![image](https://user-images.githubusercontent.com/29718932/192870701-c1a02cdb-02b6-493d-b5b3-65f23cef297a.png)
